chain/network/src/peer/peer_actor.rs

```diff
@@ -608,55 +685,24 @@ impl WriteHandler<io::Error> for PeerActor {}
 impl StreamHandler<Result<Vec<u8>, ReasonForBan>> for PeerActor {
     #[perf]
     fn handle(&mut self, msg: Result<Vec<u8>, ReasonForBan>, ctx: &mut Self::Context) {
-        let msg = match msg {
-            Ok(msg) => msg,
-            Err(ban_reason) => {
-                self.ban_peer(ctx, ban_reason);
-                return;
-            }
+        let msg = if let Some(msg) = self.check_if_decoder_banned_peer(msg, ctx) {
+            msg
+        } else {
+            return;
         };
         // TODO(#5155) We should change our code to track size of messages received from Peer
         // as long as it travels to PeerManager, etc.
 
-        metrics::PEER_DATA_RECEIVED_BYTES.inc_by(msg.len() as u64);
-        metrics::PEER_MESSAGE_RECEIVED_TOTAL.inc();
+        self.update_stats_on_receiving_message(msg.len());
 
-        self.tracker.increment_received(msg.len() as u64);
-        if codec::is_forward_tx(&msg).unwrap_or(false) {
-            let r = self.txns_since_last_block.load(Ordering::Acquire);
-            if r > MAX_TXNS_PER_BLOCK_MESSAGE {
-                return;
-            }
+        if self.should_we_drop_msg_without_decoding(&msg) {
+            return;
         }
-        let mut peer_msg = match bytes_to_peer_message(&msg) {
-            Ok(peer_msg) => peer_msg,
-            Err(err) => {
-                if let Some(version) = err
-                    .get_ref()
-                    .and_then(|err| err.downcast_ref::<HandshakeFailureReason>())
-                    .and_then(|inner| {
-                        if let HandshakeFailureReason::ProtocolVersionMismatch { version, .. } =
-                            *inner
-                        {
-                            Some(version)
-                        } else {
-                            None
-                        }
-                    })
-                {
-                    debug!(target: "network", "Received connection from node with unsupported version: {}", version);
-                    self.send_message(&PeerMessage::HandshakeFailure(
-                        self.node_info.clone(),
-                        HandshakeFailureReason::ProtocolVersionMismatch {
-                            version: PROTOCOL_VERSION,
-                            oldest_supported_version: OLDEST_BACKWARD_COMPATIBLE_PROTOCOL_VERSION,
-                        },
-                    ));
-                } else {
-                    info!(target: "network", "Received invalid data {:?} from {}: {}", logging::pretty_vec(&msg), self.peer_info, err);
-                }
-                return;
-            }
+
+        let mut peer_msg = if let Some(peer_msg) = self.try_to_decode_to_peer_message(&msg) {
+            peer_msg
+        } else {
+            return;
         };
```

Comment:
```suggestion
        let mut peer_msg = match self.try_to_decode_to_peer_message(&msg) {
            Some(peer_msg) => peer_msg,
            None => return,
        };
```

One line shorter and, imo, neater. My rule of thumb: every "if let else" is better written as match. Soonish we'll be able to write `let Some(mut peer_msg) = self.try_to_decode_to_peer_message(&msg)  else { return };`

Reply:
I agree. I wish there was an even shorted method.
 It's too bad that we can write
```
let mut peer_msg = self.try_to_decode_to_peer_message(&msg)?;
```
where `try_to_decode_to_peer_message` returns `Option<...>`

Reply:
Rust `?` works only with, `Result`, right?

Reply:
>Rust ? works only with, Result, right?

No, it also works with `Option` and [`ControlFlow`](https://doc.rust-lang.org/stable/std/ops/enum.ControlFlow.html), but it can't interconvert between diffrent container types:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d5cfd40ecea9d548c5d21cab1b14a466